### PR TITLE
Add `torch.nested` namespace

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
First step towards #83775
- only `to_padded_tensor` is moved to the nested namespace for now
- following the schema used for `special`, `fft`, `linalg` and other namespaces, nested functions are registered in native_functions.yaml as `nested_{function_name}` and are bound to the desired Python name in
`torch/nested/__init__.py`, and the desired C++ name in `torch/csrc/api/include/torch/nested.h`.

~~**Question**: should we keep the documentation for `Tensor.to_padded_tensor` or can this deleted since it is shared by `torch.nested.to_padded_tensor`?~~

[generated nested docs](https://docs-preview.pytorch.org/84102/nested.html?highlight=nested#module-torch.nested)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84102

Differential Revision: [D39361148](https://our.internmc.facebook.com/intern/diff/D39361148)

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -12609,6 +12609,17 @@
...
   python_module: linalg
 
+## Functions related to the `torch.nested` namespace
+# Note [nested namespace binding]
+# Functions in the nested python module should have their names start with
+#   "nested_" underscore and be bound to the desired Python name in
+#   torch/nested/__init__.py, and the desired C++ name in torch/csrc/api/include/torch/nested.h.
+#   The "nested_" names should be hidden from the user and not documented.
+
+- func: nested_to_padded_tensor(Tensor self, float padding, int[]? output_size=None) -> Tensor
```
Comment (drisspg) [inline aten/src/ATen/native/native_functions.yaml:12619]: Is it the pattern to create aliases for namespaced functions? I think it makes more sense to only have one function. If the pattern is to create aliases for namespace functions then then we should create a name neseted_aliases.cpp and move the definitions there.

File: aten/src/ATen/native/nested/NestedTensorMath.cpp [hunk 2]
```diff
@@ -414,6 +414,14 @@ Tensor NestedTensor_to_padded_tensor_generic(
   return ret_val;
 }
 
+// alias for to_padded_tensor in nested namespace
```
Comment (drisspg) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:417]: Going to go and refactor the nestedtensor math file and split to smaller more logical cpp files. Hence the comment above

File: caffe2/CMakeLists.txt [hunk 3]
```diff
@@ -398,6 +398,7 @@ if(NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
     "${TORCH_SRC_DIR}/csrc/autograd/generated/python_nn_functions.cpp"
     "${TORCH_SRC_DIR}/csrc/autograd/generated/python_fft_functions.cpp"
     "${TORCH_SRC_DIR}/csrc/autograd/generated/python_linalg_functions.cpp"
+    "${TORCH_SRC_DIR}/csrc/autograd/generated/python_nested_functions.cpp"
```
Comment (drisspg) [inline caffe2/CMakeLists.txt:401]: Does this have internal impacts that need to be replicated in bzl files?

File: docs/source/nested.rst [hunk 4]
```diff
@@ -142,14 +142,11 @@ nested_tensor([
 
 Note that nt.unbind()[0] is not a, but rather a slice of the underlying memory, which represents the first entry or constituent of the NestedTensor.
 
-Nested tensor methods
+Nested tensor functions
 +++++++++++++++++++++++++
 
-The following Tensor methods are related to nested tensors:
+The following functions are related to nested tensors:
```
Comment (drisspg) [inline docs/source/nested.rst:148]: "The following functions are specific to nested tensors:" ?

File: test/cpp/api/CMakeLists.txt [hunk 5]
```diff
@@ -18,6 +18,7 @@ set(TORCH_API_TEST_SOURCES
   ${TORCH_API_TEST_DIR}/moduledict.cpp
   ${TORCH_API_TEST_DIR}/modulelist.cpp
   ${TORCH_API_TEST_DIR}/modules.cpp
+  ${TORCH_API_TEST_DIR}/nested.cpp
```
Comment (drisspg) [inline test/cpp/api/CMakeLists.txt:21]: same bazel again here

File: test/test_nestedtensor.py [hunk 6]
```diff
@@ -274,6 +274,13 @@ def test_to_padded_tensor_on_empty_tensor(self):
         nt = torch.nested_tensor([])
         empty = nt.to_padded_tensor(4)
         self.assertEqual(empty, torch.tensor([]))
+
+    def test_nested_namespace(self):
+        nt = torch.nested_tensor([torch.randn(2, 3), torch.randn(4, 5)])
+        result = nt.to_padded_tensor(4)
```
Comment (drisspg) [inline test/test_nestedtensor.py:280]: yeah I hope it isn't the pattern duplicate torch.* namespace functions into the aliases because having both feels really wrong

File: torch/csrc/jit/passes/normalize_ops.cpp [hunk 7]
```diff
@@ -158,7 +159,8 @@ const std::unordered_map<Symbol, Symbol>& getOperatorAliasMap() {
       {aten::special_softmax, aten::softmax},
       {aten::special_gammainc, aten::igamma},
       {aten::special_gammaincc, aten::igammac},
-      {aten::special_gammaln, aten::lgamma}};
+      {aten::special_gammaln, aten::lgamma},
+      {aten::nested_to_padded_tensor, aten::to_padded_tensor}};
```
Comment (drisspg) [inline torch/csrc/jit/passes/normalize_ops.cpp:163]: I guess we do the aliasing thing..

File: docs/source/nested.rst (hunk 4 above)
Comment (mikaylagawarecki) [reply docs/source/nested.rst:148]: Hm, so these will be functions in the nested namespace, hence the change, but they still exist as methods. What do you think would be the best way of putting this across?

File: caffe2/CMakeLists.txt (hunk 3 above)
Comment (mikaylagawarecki) [reply caffe2/CMakeLists.txt:401]: Good catch, I probably need to import this and make those additions.

File: docs/source/nested.rst (hunk 4 above)
Comment (albanD) [reply docs/source/nested.rst:148]: I think that only functions should be here and methods will be documented on the Tensor object directly?

File: test/allowlist_for_publicAPI.json [hunk 8]
```diff
@@ -1225,6 +1225,9 @@
   "torch.multiprocessing.spawn": [
     "Optional"
   ],
+  "torch.nested": [
```
Comment (albanD) [inline test/allowlist_for_publicAPI.json:1228]: Why do you need this since you already have a `__all__` there?

File: test/test_nestedtensor.py (hunk 6 above)
Comment (albanD) [reply test/test_nestedtensor.py:280]: This one is method-only. So there is nothing in `torch.*`.

File: torch/_tensor_docs.py [hunk 9]
```diff
@@ -6511,55 +6511,3 @@ def callable(a, b) -> number
 
 """,
 )
-
-add_docstr_all(
```
Comment (albanD) [inline torch/_tensor_docs.py:6515]: You want to keep an entry for this method.
It can be simply `See torch.nested.to_padded_tensor()` (like many others in this file that point to torch.)

**Review by albanD — COMMENTED** [review]
Sounds good!

File: test/allowlist_for_publicAPI.json (hunk 8 above)
Comment (mikaylagawarecki) [reply test/allowlist_for_publicAPI.json:1228]: TestCorrectModuleNames seemed to fail with just the `__all__`  before adding it here. I'll check again to make sure

File: test/test_nestedtensor.py (hunk 6 above)
Comment (drisspg) [reply test/test_nestedtensor.py:280]: Ohh see when then added a function that goes in nested namespace? Do we need that? Should only functional variants go in nested namespace?

Comment (albanD) [reply test/test_nestedtensor.py:280]: In general yes because you don't want nested-specific methods on plain Tensor.
But I guess it's too late :p

File: test/allowlist_for_publicAPI.json (hunk 8 above)
Comment (mikaylagawarecki) [reply test/allowlist_for_publicAPI.json:1228]: Double checked this, test_public_bindings.py:test_correct_module_names seems to fail without this

File: docs/source/nested.rst (hunk 4 above)
Comment (mikaylagawarecki) [reply docs/source/nested.rst:148]: Just to confirm, we still want to_padded_tensor to live as a function in the torch.nested namespace, is that correct?

File: aten/src/ATen/native/nested/NestedTensorMath.cpp (hunk 2 above)
Comment (mikaylagawarecki) [reply aten/src/ATen/native/nested/NestedTensorMath.cpp]: Added NestedTensorAliases.cpp, will import and include in the relevant bzl files

File: docs/source/nested.rst (hunk 4 above)
Comment (drisspg) [reply docs/source/nested.rst:148]: Alban suggested to me that we could potentially remove the method on the tensor base. And use the top level python api to create the method. This python level implementation would under the hood call the torch.nested.to_padded_tensor() function

But yeah I think as long as there is one **function** exposed in python not having both `torch.to_padded_tensor(nt, 0)` and `torch.nested.to_padded_tensor(nt, 0 )` that is good.

**Review by drisspg — COMMENTED** [review]
One more comment, If the native function `to_padded_tensor` didn't exist and we were adding it to native functions.yaml would we only have this entry? : 
```
- func: nested_to_padded_tensor(Tensor self, float padding, int[]? output_size=None) -> Tensor
  python_module: nested
  variants: function
``` 

If so I think that we should consider removing [to_padded_tensor](https://github.com/pytorch/pytorch/blob/85931eaa6beab53d138f873d3505aee34e98ee89/aten/src/ATen/native/native_functions.yaml#L13073), which defines a method entirely. 

Is there a great reason to have both of these native_function entries, maybe I am missing something?

**Comment by mikaylagawarecki** [comment]
@drisspg  I'm slightly concerned with breaking BC by removing the method since we had already unprivated and documented `Tensor.to_padded_tensor` :(

**Comment by drisspg** [comment]
> @drisspg I'm slightly concerned with breaking BC by removing the method since we had already unprivated and documented `Tensor.to_padded_tensor` :(

Yeah but we aren't even prototype lol and no one really uses nested tensor. I think it would be okay. @albanD Thoughts?

Would an alternative if we still wanted to be a method in python be adding it as a method to this class:https://github.com/pytorch/pytorch/blob/65f98eb47dbf75335d08f7676835a5e1f1fc3574/torch/_tensor.py#L99

**Comment by mikaylagawarecki** [comment]
git rebase -s

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39361148).

**Comment by mikaylagawarecki** [comment]
Removed `Tensor.to_padded_tensor()` and imported + changed internal invocations of `Tensor.to_padded_tensor --> torch.nested.to_padded_tensor`

File: build_variables.bzl [hunk 10]
```diff
@@ -983,6 +983,7 @@ def glob_libtorch_python_sources(gencode_pattern = ":generate-code[{}]"):
         "torch/csrc/autograd/generated/python_functions_2.cpp",
         "torch/csrc/autograd/generated/python_functions_3.cpp",
         "torch/csrc/autograd/generated/python_functions_4.cpp",
+    "torch/csrc/autograd/generated/python_nested_functions.cpp",
```
Comment (drisspg) [inline build_variables.bzl:986]: nit: is spacing weird on this?

**Review by drisspg — APPROVED** [review]
Awesome great work Mikayla!

File: build_variables.bzl (hunk 10 above)
Comment (mikaylagawarecki) [reply build_variables.bzl:986]: oh shoot thanks I didn't notice this

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39361148).

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39361148).

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39361148).
